python/pyspark/pandas/groupby.py

```diff
@@ -2121,11 +2121,22 @@ def _limit(self, n: int, asc: bool) -> FrameLike:
             )
         )
 
-        sdf = (
-            sdf.withColumn(tmp_col, F.row_number().over(window))
-            .filter(F.col(tmp_col) <= n)
-            .drop(tmp_col)
-        )
+        if n >= 0 or LooseVersion(pd.__version__) < LooseVersion("1.4.0"):
+            sdf = (
+                sdf.withColumn(tmp_row_num_col, F.row_number().over(window))
+                .filter(F.col(tmp_row_num_col) <= n)
+                .drop(tmp_row_num_col)
+            )
+        else:
+            # Pandas supports Groupby positional indexing since v1.4.0
+            # https://pandas.pydata.org/docs/whatsnew/v1.4.0.html#groupby-positional-indexing
+            tmp_cnt_col = verify_temp_column_name(sdf, "__group_count__")
+            sdf = (
+                sdf.withColumn(tmp_row_num_col, F.row_number().over(window))
+                .withColumn(tmp_cnt_col, F.count("*").over(Window.partitionBy(*groupkey_scols)))
+                .filter(F.col(tmp_row_num_col) - F.col(tmp_cnt_col) <= n)
```

Comment:
The filter condition is cool but may not be easy to understand. Shall we add (an example as?) a comment?

Below is just a reference.
```py
+-----------------+--------------+---+---+-----------------+--------------+---------------+
|__index_level_0__|__groupkey_0__|  A|  B|__natural_order__|__row_number__|__group_count__|
+-----------------+--------------+---+---+-----------------+--------------+---------------+
|                0|             g|  g| g0|      17179869184|             1|              4|
|                1|             g|  g| g1|      42949672960|             2|              4|
|                2|             g|  g| g2|      60129542144|             3|              4|
|                4|             h|  h| h0|     111669149696|             1|              2|
+-----------------+--------------+---+---+-----------------+--------------+---------------+

```

Reply:
Below as a reference in case you need it.
```py
>>> sdf.withColumn(tmp_row_num_col, F.row_number().over(window)).withColumn(tmp_cnt_col, F.count("*").over(Window.partitionBy(*groupkey_scols))).show()
+-----------------+--------------+---+---+-----------------+--------------+---------------+
|__index_level_0__|__groupkey_0__|  A|  B|__natural_order__|__row_number__|__group_count__|
+-----------------+--------------+---+---+-----------------+--------------+---------------+
|                0|             g|  g| g0|      17179869184|             1|              4|
|                1|             g|  g| g1|      42949672960|             2|              4|
|                2|             g|  g| g2|      60129542144|             3|              4|
|                3|             g|  g| g3|      85899345920|             4|              4|
|                4|             h|  h| h0|     111669149696|             1|              2|
|                5|             h|  h| h1|     128849018880|             2|              2|
+-----------------+--------------+---+---+-----------------+--------------+---------------+


>>> sdf.withColumn(tmp_row_num_col, F.row_number().over(window)).withColumn(tmp_cnt_col, F.count("*").over(Window.partitionBy(*groupkey_scols))).filter(F.col(tmp_row_num_col) - F.col(tmp_cnt_col) <= -1).show()
+-----------------+--------------+---+---+-----------------+--------------+---------------+
|__index_level_0__|__groupkey_0__|  A|  B|__natural_order__|__row_number__|__group_count__|
+-----------------+--------------+---+---+-----------------+--------------+---------------+
|                0|             g|  g| g0|      17179869184|             1|              4|
|                1|             g|  g| g1|      42949672960|             2|              4|
|                2|             g|  g| g2|      60129542144|             3|              4|
|                4|             h|  h| h0|     111669149696|             1|              2|
+-----------------+--------------+---+---+-----------------+--------------+---------------+
```

Reply:
Added!